File: apps/api/app/schemas/task.py
"""Task schemas."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.task import AIStatus, TaskPriority, TaskStatus, TaskType
from app.schemas.base import TimestampSchema, UUIDSchema


class LabelBase(BaseModel):
    name: str = Field(..., max_length=100)
    color: str = Field(..., max_length=20)


class LabelCreate(LabelBase):
    project_id: str


class LabelUpdate(BaseModel):
    name: str | None = Field(None, max_length=100)
    color: str | None = Field(None, max_length=20)


class LabelResponse(LabelBase, UUIDSchema, TimestampSchema):
    project_id: str

    model_config = ConfigDict(from_attributes=True)


class TaskActivityResponse(UUIDSchema):
    task_id: str
    user_id: str
    action: str
    old_value: dict[str, Any] | None = None
    new_value: dict[str, Any] | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TaskWatcherResponse(UUIDSchema):
    task_id: str
    user_id: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TaskBase(BaseModel):
    title: str = Field(..., max_length=500)
    description: str | None = None
    status: TaskStatus = TaskStatus.TODO
    priority: TaskPriority = TaskPriority.MEDIUM
    type: TaskType = TaskType.TASK

    parent_task_id: str | None = None
    assignee_id: str | None = None
    reporter_id: str | None = None
    assigned_agent_id: str | None = None

    due_date: datetime | None = None
    start_date: datetime | None = None
    completed_at: datetime | None = None

    estimated_hours: float | None = None
    actual_hours: float | None = None
    position: float = 0.0

    metadata_: dict[str, Any] | None = Field(None, alias="metadata")
    ai_context: dict[str, Any] | None = None
    ai_status: AIStatus = AIStatus.IDLE

    is_recurring: bool = False
    recurrence_rule: str | None = Field(None, max_length=500)
    next_occurrence: datetime | None = None


class TaskCreate(TaskBase):
    project_id: str
    label_ids: list[str] = Field(default_factory=list)


class TaskUpdate(BaseModel):
    title: str | None = Field(None, max_length=500)
    description: str | None = None
    status: TaskStatus | None = None
    priority: TaskPriority | None = None
    type: TaskType | None = None

    parent_task_id: str | None = None
    assignee_id: str | None = None
    reporter_id: str | None = None
    assigned_agent_id: str | None = None

    due_date: datetime | None = None
    start_date: datetime | None = None
    completed_at: datetime | None = None

    estimated_hours: float | None = None
    actual_hours: float | None = None
    position: float | None = None

    metadata_: dict[str, Any] | None = Field(None, alias="metadata")
    ai_context: dict[str, Any] | None = None
    ai_status: AIStatus | None = None

    is_recurring: bool | None = None
    recurrence_rule: str | None = Field(None, max_length=500)
    next_occurrence: datetime | None = None

    label_ids: list[str] | None = None


class TaskResponse(TaskBase, UUIDSchema, TimestampSchema):
    project_id: str
    labels: list[LabelResponse] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class TaskBulkUpdate(BaseModel):
    task_ids: list[str]
    status: TaskStatus | None = None
    priority: TaskPriority | None = None
    assignee_id: str | None = None
    label_ids: list[str] | None = None

File: packages/types/src/task.ts
export type TaskStatus = "todo" | "in_progress" | "in_review" | "blocked" | "done" | "cancelled";

export type TaskPriority = "low" | "medium" | "high" | "urgent";

export type TaskType = "task" | "bug" | "feature" | "epic" | "story" | "subtask";

export type AIStatus = "idle" | "planning" | "running" | "waiting" | "failed" | "completed";

export interface Label {
  id: string;
  projectId: string;
  name: string;
  color: string;
  createdAt: string;
  updatedAt: string;
}

export interface TaskActivity {
  id: string;
  taskId: string;
  userId: string;
  action: string;
  oldValue: Record<string, unknown> | null;
  newValue: Record<string, unknown> | null;
  createdAt: string;
}

export interface TaskWatcher {
  id: string;
  taskId: string;
  userId: string;
  createdAt: string;
}

export interface Task {
  id: string;
  projectId: string;
  parentTaskId: string | null;
  title: string;
  description: string | null;
  status: TaskStatus;
  priority: TaskPriority;
  type: TaskType;

  assigneeId: string | null;
  reporterId: string | null;
  assignedAgentId: string | null;

  dueDate: string | null;
  startDate: string | null;
  completedAt: string | null;

  estimatedHours: number | null;
  actualHours: number | null;
  position: number;

  metadata: Record<string, unknown> | null;
  aiContext: Record<string, unknown> | null;
  aiStatus: AIStatus;

  isRecurring: boolean;
  recurrenceRule: string | null;
  nextOccurrence: string | null;

  createdAt: string;
  updatedAt: string;

  labels?: Label[];
}

export interface TaskCreateInput {
  projectId: string;
  parentTaskId?: string | null;
  title: string;
  description?: string | null;
  status?: TaskStatus;
  priority?: TaskPriority;
  type?: TaskType;

  assigneeId?: string | null;
  reporterId?: string | null;
  assignedAgentId?: string | null;

  dueDate?: string | null;
  startDate?: string | null;
  completedAt?: string | null;

  estimatedHours?: number | null;
  actualHours?: number | null;
  position?: number;

  metadata?: Record<string, unknown> | null;
  aiContext?: Record<string, unknown> | null;
  aiStatus?: AIStatus;

  isRecurring?: boolean;
  recurrenceRule?: string | null;
  nextOccurrence?: string | null;

  labelIds?: string[];
}

export interface TaskUpdateInput extends Partial<Omit<TaskCreateInput, "projectId">> {
  id: string;
}

export interface TaskBulkUpdateInput {
  taskIds: string[];
  status?: TaskStatus;
  priority?: TaskPriority;
  assigneeId?: string | null;
  labelIds?: string[];
}
